File: 155-min-stack/min-stack.java
class MinStack {
    ArrayDeque<Integer> stack;
    ArrayDeque<Integer> minStack;
    public MinStack() {
        stack = new ArrayDeque();
        minStack = new ArrayDeque();
    }
    
    public void push(int value) {
        stack.push(value);
        if (minStack.isEmpty()) {
            minStack.push(value);
        }
        else {
            minStack.push(Math.min(value, minStack.peek()));
        }
    }
    
    public void pop() {
        stack.pop();
        minStack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(value);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155-min-stack/min-stack.py
class MinStack:
    def __init__(self):
        self.st = []
        self.mnStack = []
    def push(self, value: int) -> None:
        self.st.append(value)
        if not self.mnStack:
            self.mnStack.append(value)
        else:
            self.mnStack.append(min(value, self.mnStack[-1]))
    def pop(self) -> None:
        self.mnStack.pop()
        return self.st.pop()
    def top(self) -> int:
        return self.st[-1]
    def getMin(self) -> int: # O(1) constant time
        return self.mnStack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(value)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()